File: scratch/pull_and_redeploy.js
// pull_and_redeploy.js — uses ssh2 npm package to connect to Vultr and redeploy
// Run: node scratch/pull_and_redeploy.js

const { Client } = require('ssh2');
const path = require('path');

const HOST     = '65.20.85.75';
const PORT     = 22;
const USERNAME = 'graceandforce';
const PASSWORD = 'wvpi2!ZnTcV];ncy';
const APP_DIR  = '/home/graceandforce/debate-engine';

const NVM = 'export NVM_DIR="$HOME/.nvm"; [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"';

const COMMANDS = [
  { label: 'git pull', cmd: `cd ${APP_DIR} && git pull origin main` },
  { label: 'npm install (backend)', cmd: `${NVM}; cd ${APP_DIR}/backend && npm install --production 2>&1 | tail -5` },
  { label: 'pm2 restart', cmd: `${NVM}; pm2 restart grace-api` },
  { label: 'npm install (frontend)', cmd: `${NVM}; cd ${APP_DIR}/frontend && npm install 2>&1 | tail -5` },
  { label: 'build frontend', cmd: `${NVM}; cd ${APP_DIR}/frontend && npm run build 2>&1 | tail -10` },
  { label: 'deploy frontend', cmd: `echo 'wvpi2!ZnTcV];ncy' | sudo -S rsync -a --delete ${APP_DIR}/frontend/dist/ /var/www/grace-and-force/frontend/` },
  { label: 'pm2 status',  cmd: `${NVM}; pm2 list` },
  {
    label: 'smoke test /api/evaluate',
    cmd: `curl -s -X POST http://localhost:5000/api/evaluate \
-H "Content-Type: application/json" \
-d '{"transcript":[{"role":"agent","text":"Do you think AI is beneficial?"},{"role":"user","text":"Yes I strongly believe AI is beneficial because it helps people work faster and solve complex problems with data."}],"topic":"AI is beneficial","isJunior":false,"studentId":"test","name":"Test","classLevel":"Level 3"}' \
| python3 -c "import sys,json; r=json.load(sys.stdin); print('SCORE:', r.get('overall_score','ERR'), '| ERROR:', r.get('error','none'))"`,
  },
];

async function runCommand(conn, cmd) {
  return new Promise((resolve, reject) => {
    conn.exec(cmd, (err, stream) => {
      if (err) return reject(err);
      let out = '';
      stream.on('data', d => { out += d; });
      stream.stderr.on('data', d => { out += d; });
      stream.on('close', () => resolve(out.trim()));
    });
  });
}

async function deploy() {
  const conn = new Client();

  await new Promise((resolve, reject) => {
    conn.on('ready', resolve);
    conn.on('error', reject);
    conn.connect({ host: HOST, port: PORT, username: USERNAME, password: PASSWORD, readyTimeout: 15000 });
  });

  console.log(`✅ Connected to ${HOST}\n`);

  for (const { label, cmd } of COMMANDS) {
    console.log(`\n>>> ${label}`);
    try {
      const output = await runCommand(conn, cmd);
      console.log(output || '(no output)');
    } catch (e) {
      console.log('ERROR:', e.message);
    }
  }

  conn.end();
  console.log('\n✅ Redeploy complete!');
}

deploy().catch(e => { console.error('Fatal:', e.message); process.exit(1); });

File: scratch/pull_and_redeploy.py
import paramiko

HOST     = "65.20.85.75"
PORT     = 22
USERNAME = "graceandforce"
PASSWORD = "wvpi2!ZnTcV];ncy"
APP_DIR  = "/home/graceandforce/debate-engine"

def run(client, cmd, label=""):
    print(f"\n>>> {label or cmd}")
    stdin, stdout, stderr = client.exec_command(cmd)
    out = stdout.read().decode("utf-8", "ignore").strip()
    err = stderr.read().decode("utf-8", "ignore").strip()
    if out:
        print(out)
    if err:
        print("STDERR:", err)
    return out

def deploy():
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

    print(f"Connecting to {HOST}...")
    client.connect(HOST, PORT, USERNAME, PASSWORD, timeout=20)
    print("Connected!\n")

    nvm = 'export NVM_DIR="$HOME/.nvm"; [ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"'

    # 1. Git pull
    run(client, f"cd {APP_DIR} && git pull origin main", "git pull")

    # 2. Install any new npm deps (quick, skips if nothing changed)
    run(client, f'{nvm}; cd {APP_DIR}/backend && npm install --production', "npm install")

    # 3. Restart PM2
    run(client, f'{nvm}; pm2 restart grace-api', "pm2 restart")

    # 4. Status
    run(client, f'{nvm}; pm2 status', "pm2 status")

    # 5. Quick smoke test of evaluate endpoint
    run(client,
        'curl -s -X POST http://localhost:5000/api/evaluate '
        '-H "Content-Type: application/json" '
        '-d \'{"transcript":[{"role":"agent","text":"Do you agree?"},{"role":"user","text":"Yes I strongly agree because social media connects people and helps share ideas globally."}],'
        '"topic":"Social media","isJunior":false,"studentId":"test","name":"Test","classLevel":"Level 3"}\' '
        '| python3 -c "import sys,json; r=json.load(sys.stdin); print(\'OK - Score:\', r.get(\'overall_score\', r.get(\'error\',\'??\')))"',
        "smoke test evaluate endpoint"
    )

    client.close()
    print("\nDone!")

if __name__ == "__main__":
    deploy()
